index.js

```diff
@@ -1,85 +1,157 @@
- * a single discovery-swarm instance.
+ * The Dat object. Manages multiple archives in
+ * a single discovery-swarm-web instance.
  * @param {Object} opts   Default options to use for the dat.
  */
 class Dat extends EventEmitter {
   constructor (opts) {
     super()
     this.opts = opts || {}
 
-    if (!this.opts.id) this.opts.id = randomBytes(32)
+    if (!this.opts.id) this.opts.id = crypto.randomBytes(32)
 
-    this.repos = []
+    this.archives = []
+
+    const discoveryOpts = Object.assign({
+      stream: (info) => this._replicate(info)
+    }, this.opts)
+    this.swarm = new DiscoverySwarmWeb(discoveryOpts)
+
+    const storageOpts = Object.assign({
+      name: STORAGE_NAME,
+      storeName: STORAGE_COLLECTION_NAME
+    }, this.opts)
+    this.persistence = RAW(storageOpts)
   }
 
   /**
-   * Returns a repo with the given url. Returns undefined
-   * if no repository is found with that url.
-   * @param  {url} url      The url of the repo.
-   * @return {Repo|undefined}  The repo object with the corresponding url.
+   * Returns a archive with the given url.
+   * @param  {url} url      The url of the archive.
+   * @return {hyperdrive}  The archive object with the corresponding url.
    */
   get (url, opts) {
-    const repo = this.repos.find((repo) => repo.url === url)
-    if (repo) return repo
+    const archive = this.archives.find((archive) => archive.url === url)
+    if (archive) return archive
     return this._add(url, opts)
   }
 
   _add (url, opts) {
     if (this.destroyed) throw new Error('client is destroyed')
     if (!opts) opts = {}
+    const finalOpts = Object.assign({}, this.opts, opts)
 
     let key = null
 
     if (url) key = encoding.decode(url)
 
-    const repo = new Repo(key, xtend(this.opts, opts))
-    this.repos.push(repo)
+    // If no key was provided, generate one
+    if (!key) {
+      const keyPair = crypto.keyPair()
+      key = keyPair.publicKey
+      finalOpts.secretKey = keyPair.secretKey
+    }
+
+    const keyString = encoding.encode(key)
+
+    const rawStorage = finalOpts.db || (finalOpts.persist ? this.persistence : RAM)
+    const storage = (file) => {
+      return rawStorage(keyString + '/' + file)
+    }
+    const archive = hyperdrive(storage, key, Object.assign({
+      sparse: true
+    }, finalOpts))
+
+    archive.url = DAT_PROTOCOL + keyString
 
-    repo.ready(() => {
-      this.emit('repo', repo)
+    this.archives.push(archive)
+
+    archive.ready(() => {
+      this.swarm.join(archive.discoveryKey)
+      this.emit('archive', archive)
     })
 
-    return repo
+    return archive
   }
 
   create (opts) {
     return this._add(null, opts)
   }
 
   has (url) {
-    return !!this.repos.find((repo) => repo.url === url)
+    return !!this.archives.find((archive) => archive.url === url)
+  }
+
+  _replicate (info) {
+    var stream = hypercoreProtocol({
+      id: this.opts.id,
+      live: true,
+      encrypt: true,
+      extensions: ['ephemeral', 'session-data']
```

Comment:
Can you give me a tldr on how we're using these extensions? (not sure how many people are using these yet so could be a good example use case)

Reply:
They're not being used, I copied this part from beaker. I want to eventually support them, though. I'll remove it for now.